Lilu/Headers/kern_util.hpp

```diff
@@ -944,4 +945,279 @@ class evector_base {
 template <typename T, void (*deleter)(T)=emptyDeleter<T>>
 class evector : public evector_base<typename remove_reference<T>::type, T, deleter> { };
 
+/**
+ *  Represents a circular buffer protected by a recursive mutex lock
+ */
+template <typename T>
+struct CircularBuffer {
+private:
+	/**
+	 *  The internal storage
+	 */
+	T *storage {nullptr};
+	
+	/**
+	 *  The buffer capacity
+	 */
+	IOItemCount size {0};
+	
+	/**
+	 *  The current index for the next read operation
+	 */
+	IOItemCount indexr {0};
+	
+	/**
+	 *  The current index for the next write operation
+	 */
+	IOItemCount indexw {0};
+	
+	/**
+	 *  The current number of elements in the buffer
+	 */
+	IOItemCount count {0};
+	
+	/**
+	 *  The recursive mutex lock that protects the buffer
+	 */
+	IORecursiveLock *lock {nullptr};
+	
+public:
+	/**
+	 *  Initialize a circular buffer
+	 *
+	 *  @param buffer A non-null storage buffer
+	 *  @param capacity The total number of elements
+	 *  @return `true` on success, `false` otherwise.
+	 *  @warning The caller is responsbile for managing the lifecycle of the given storage buffer.
+	 */
+	bool init(T *buffer, IOItemCount capacity) {
+		storage = buffer;
+		size = capacity;
+		indexr = 0;
```

Comment:
Is there really a need to initialise these two? In general simple to initialise fields is easy to miss, so we make sure all fields have inline init, and only special ones get direct init.

Reply:
Forgot to remove them. Done.